### ISSUE TITLE
Simulation halt when resupply missions arrive 

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#719]
**Describe the bug**
- At sol 1:100, the simulation appears **frozen** while waiting for all resupply missions to arrive at each settlement.

**Expected and unexpected behavior**
- If there are only resources (amount resource, item resource, equipment) in a resupply mission, the simulation does not need to be on pause.
- However, if some buildings come with a resupply mission, the clock pulse should stop and all the path finding methods or tasks need to be cancelled, or else it would create exceptions, owning to new buildings being added to a settlement.

**Java classes involved**
So, it has nothing to do with the clock since `TransportManager` is on its own thread and somehow it halts the clock pulse.

**Screenshots**
![image](https://user-images.githubusercontent.com/1168584/194717419-e20765c1-56ae-4f11-8c1c-789f03582695.png)

**Stacktrace**
```
01-Adir-01:100.109 (Info) TransportManager : Transport item arrived at 01-Adir-01 : Tian Cheng
01-Adir-01:100.109 (Info) Resupply : [Europa] Resuply has arrived
01-Adir-01:100.109 (Info) TransportManager : Transport item arrived at 01-Adir-01 : Europa
01-Adir-01:100.109 (Info) Resupply : [New Plymouth] Resuply has arrived
01-Adir-01:100.109 (Info) TransportManager : Transport item arrived at 01-Adir-01 : New Plymouth
01-Adir-01:100.109 (Info) Resupply : [New Seoul] Resuply has arrived
01-Adir-01:100.109 (Info) TransportManager : Transport item arrived at 01-Adir-01 : New Seoul
01-Adir-01:100.109 (Info) Resupply : [Starcity] Resuply has arrived
01-Adir-01:100.109 (Info) TransportManager : Transport item arrived at 01-Adir-01 : Starcity
01-Adir-01:100.109 (Info) Resupply : [Zvezda] Resuply has arrived
01-Adir-01:100.109 (Warning) MicroInventory : [Zvezda] Storage is full. Excess 500.0 kg food.
01-Adir-01:100.109 (Info) TransportManager : Transport item arrived at 01-Adir-01 : Zvezda
01-Adir-01:100.109 (Info) Resupply : [Kyocera] Resuply has arrived
01-Adir-01:100.109 (Info) TransportManager : Transport item arrived at 01-Adir-01 : Kyocera
```

**Specifications  (please complete the following information):**
 - pre-3.4.0 build 7471

**Additional context**
- If I hit *pause* before all the transport items are processed, it will still continue to process them until they are done.
- Questions: What causes the simulation to be put on pause ?  Is this *undesirable* effect actually something that we want to keep and leave it the way it behaves ? 
- See https://github.com/mars-sim/mars-sim/commit/884a5fd353725f55bfc1bbde33e8ca5af174aac6 for related issue

**Comment by bevans2000** [comment]
The excessive events being generated were the reason I put #309 on hold. The root cause of this problem is we update the UI directly from the simulation. #309 breaks this coupling.
I think the Resupply missions are recent additions?

**Comment by mokun** [comment]
> The excessive events being generated were the reason I put #309 on hold. The root cause of this problem is we update the UI directly from the simulation. #309 breaks this coupling. I think the Resupply missions are recent additions?

We have resupply missions a long time ago but it happens only once an orbit. 

So we never tested and get to know the freeze until now.  

Recently I did add a brand new *Standard Resupply Mission 1* to arrive on sol1:100. 

That's why I notice this now.

**Comment by bevans2000** [comment]
It doesn't always freeze. I think you have to have the Monitor Tool open for it to slow down. Just done some tests and they worked without any slow down.

**Comment by mokun** [comment]
> It doesn't always freeze.

It just looks like the clock was put on pause because the timestamp halts at sol1:100. 

The transport manager pulse is still running.

After it was done, everything does return to normal.

**Comment by bevans2000** [comment]
Last night I did quite a few runs where there was no freezing. It depends on what state the UI is.

**Comment by bevans2000** [comment]
Of course a big change is to remove Equipment from the Unit hierarchy. We've discussed this but the simulation has 100s of Equipment and they don't require the same level of details that normal Units have. So when adding extra Equipment in a Resupply mission it would not trigger the NewUnit events that appear to be the starting point of the slowdown.
Robot would need to be moved to the directly under Unit which again I think is the correct model.
If we want to do this; we should create a Project and sketch out some steps to achieve the objective.

**Comment by mokun** [comment]
If I happen to open any pie or bar charts in the Monitor Tool while it's delivering a resupply, the following will result:

```
01-Adir-01:100.078 (Info) Resupply : [Schiaparelli Point] Resuply has arrived
01-Adir-01:100.078 (Info) TransportManager : Transport item arrived at 01-Adir-01 : Schiaparelli Point
01-Adir-01:100.078 (Info) Resupply : [New Pompeii] Resuply has arrived
01-Adir-01:100.078 (Severe) MasterClock : Can't send out clock pulse: 
java.util.ConcurrentModificationException
	at java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1511)
	at java.base/java.util.HashMap$KeyIterator.next(HashMap.java:1534)
	at org.mars_sim.msp.core.UnitManager.fireUnitManagerUpdate(UnitManager.java:591)
	at org.mars_sim.msp.core.UnitManager.addUnit(UnitManager.java:286)
	at org.mars_sim.msp.core.equipment.EquipmentFactory.createEquipment(EquipmentFactory.java:62)
	at org.mars_sim.msp.core.equipment.EquipmentFactory.createEquipment(EquipmentFactory.java:81)
	at org.mars_sim.msp.core.interplanetary.transport.resupply.Resupply.deliverOthers(Resupply.java:460)
	at org.mars_sim.msp.core.interplanetary.transport.resupply.Resupply.performArrival(Resupply.java:188)
	at org.mars_sim.msp.core.interplanetary.transport.TransportManager.timePassing(TransportManager.java:186)
	at org.mars_sim.msp.core.Simulation.clockPulse(Simulation.java:1273)
	at org.mars_sim.msp.core.time.MasterClock$ClockListenerTask.call(MasterClock.java:640)
	at org.mars_sim.msp.core.time.MasterClock$ClockListenerTask.call(MasterClock.java:1)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

```

**Comment by mokun** [comment]
> the simulation has 100s of Equipment and they don't require the same level of details that normal Units have. So when adding extra Equipment in a Resupply mission it would not trigger the NewUnit events that appear to be the starting point of the slowdown.

So is the culprit the UI (for instance, the Monitor Tool) ? 

When calling createEquipment() in EquipmentFactory to create that equipment instance, it creates a new unit in UnitManager. Then it calls fireUnitManagerUpdate(), which in terms fire all the listeners in the Monitor Tool ...

![image](https://user-images.githubusercontent.com/1168584/194777337-f7bd3680-df17-4da6-aea4-af455719a180.png)

In `UnitManger`, we have the following : 

```
	/**
	 * Fire a unit update event.
	 *
	 * @param eventType the event type.
	 * @param unit      the unit causing the event.
	 */
	public final void fireUnitManagerUpdate(UnitManagerEventType eventType, Unit unit) {
		if (listeners == null) {
			listeners = new HashSet<>();
		}
		synchronized (listeners) {
			for (UnitManagerListener listener : listeners) {
				listener.unitManagerUpdate(new UnitManagerEvent(this, eventType, unit));
			}
		}
	}
```

**Comment by bevans2000** [comment]
> > the simulation has 100s of Equipment and they don't require the same level of details that normal Units have. So when adding extra Equipment in a Resupply mission it would not trigger the NewUnit events that appear to be the starting point of the slowdown.
> 
> So is the culprit the UI thread used by all those TableModel classes in the Monitor Tool ?
> 
> When calling createEquipment() in EquipmentFactory to create that equipment instance, it creates a new unit in UnitManager. Then it calls fireUnitManagerUpdate(), which in terms fire all the listeners in the Monitor Tool .

Yes I believe it's part of the problem but the event is needed to add rows to the monitor tab.

**Comment by bevans2000** [comment]
> If I happen to open any pie or bar charts in the Monitor Tool while it's delivering a resupply, the following will result:
> 
That's an odd one as I don't see where the hash map is coming from. I'll have a look at this.
What did you have showing in the pie chart when it failed?
Short term, just avoid the pie chart for now.

**Comment by mokun** [comment]
> What did you have showing in the pie chart when it failed?

![image](https://user-images.githubusercontent.com/1168584/194778603-106ead99-bc8e-4022-ba5c-af0af90794b7.png)

It's when I clicked on either the first button (the pie icon) or the second button (the bar icon). Or I may have also clicked on the OK button on the pop-up panel. 

Btw, it almost sounds odd that there's only *one and only one* pie chart available for the Mars tab and it has to do with the type of Malfunction. 

But it's actually right. 

The rest of the columns on the table are just numbers and there's no way to present any other columns in a pie chart.

**Comment by bevans2000** [comment]
Well well well what a  discovery. The problem has nothing to do with the chart directly, (I think).
The problem is coming the listeners Set in ```UnitManager``` is unprotected. When firing the event for a new Unit, the Monitor Window is adding a new listeners from the UI thread hence generating a Concurrent Mod exception.
I will commit a fix to protect the listeners.

However, the real problem is the MonitorWindow adds a new listener whenever a tab is selected but never removes them. So by just switching between tabs I managed to have 35+ listeners. This could explain this freeze problem because each new Unit is triggering many listeners. This needs a bit of thought on how to fix it.

**Comment by bevans2000** [comment]
Yes that would make sense. I don't know why we recreate the tab, just needs a reset

**Comment by bevans2000** [comment]
> Yes that would make sense. I don't know why we recreate the tab, just needs a reset

I plan to have a look this evening. Also I think we can be more efficient with the UnitManagerEvent logic as well by partitioning the listeners according the the UnitType. From a quick review yesterday, the listener seem to be only interested in a particular UnitType BUT get notified of any change, i.e. adding new Equipment will trigger the listeners in all tabs but they filter out those of no interest.

**Comment by bevans2000** [comment]
Just an update on this. The ```MonitorWindow``` and associated tab models have quite a few problems in their interactions consuming a lot of processing when a UnitEvent is received. They are very wasteful and in cases keep registering the same listener multiple times.
I think I'm over the hump now. If possible can you avoid changing anything in the MonitorWindow until I merge my branch.

**Comment by mokun** [comment]
> avoid changing anything in the MonitorWindow until I merge my branch.

Sure. 

I did touch `MonitorWindow` yesterday as I was fixing some quirky things in `ArrivingSettlement` and the errors in two of the settlement template xml files (Phase 2-MS and Phase 3-MS), because new settlements were not able to show up in the settlement combo box and the Mars tab and settlement tab.

**Comment by bevans2000** [comment]
> > avoid changing anything in the MonitorWindow until I merge my branch.
> 
> Sure.
> 
> I did touch `MonitorWindow` yesterday as I was fixing some quirky things in `ArrivingSettlement` and the errors in two of the settlement template xml files (Phase 2-MS and Phase 3-MS), because new settlements were not able to show up in the settlement combo box and the Mars tab and settlement tab.

I found the same bug myself when reworking the ```UnitManagerEvent``` event processing. No problem, I'll merge them together.

**Comment by bevans2000** [comment]
Right, it's been a journey but I think I'm there now. Made some big but good changes to the ```MonitorWindow``` area in the latest commit.
The key recent concept is using a new ```EntityTableModel``` which takes care of the table model responsibilities. This is a templated class to ensure typesafe. The subclasses now acts as an adapter from the Entity being displayed to the underlying column model. This means that the EntityTableModel can provide a value cache for specific columns that are expensive to calculate. This make a big difference when sorting a derived/calculated column; sorting is much quicker because the values are coming out of the cache and not the backing Unit class. The trade table is a good example.
Seems to be working fine but I will try and run it some more tomorrow and then create a pull request.
However based on #723 the values are shown raw as defined in the XML files.

**Comment by mokun** [comment]
I see quite a lot of this log on a vehicle.



```
01-Adir-02:788.701 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:788.701 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:788.846 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:788.846 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:788.991 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:788.991 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.136 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.136 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.292 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.292 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.443 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.443 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.602 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.602 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.753 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.753 (Warning) EntityTableModel : Entity not in model : Martian Escape
01-Adir-02:789.900 (Warning) EntityTableModel : Entity not in model : Martian Escape
```

It came from entityValueUpdated() in EntityTableModel.

```
    /**
     * A range of column values have changed.
     * @param entity
     * @param firstCol
     * @param lastCol
     */
	protected void entityValueUpdated(T entity, int firstCol, int lastCol) {
        int rowIndex = entities.indexOf(entity);
        if (rowIndex < 0) {
            logger.warning("Entity not in model : " + entity.toString());
            return;
        }

        for(int i = firstCol; i<= lastCol; i++) {
            if (cachedColumns.contains(i)) {
                setCacheValue(entity, i, null); // Clear cached value
            }
            fireTableCellUpdated(rowIndex, i);
        }
    }
```

Can you take a quick look ?

**Comment by mokun** [comment]
Another minor thing is that I've wanted to see proper column width to be computed. 

In `TableTab`, we have adjustColumnWidth() as follows. 

```
	public void adjustColumnWidth(JTable table) {
		// Gets max width for cells in column as the preferred width
		TableColumnModel columnModel = table.getColumnModel();
		for (int col = 0; col < table.getColumnCount(); col++) {
			TableColumn tableColumn = columnModel.getColumn(col);
		    int preferredWidth = tableColumn.getMinWidth() + 15;
			int w = 50;
		    TableCellRenderer rend = table.getTableHeader().getDefaultRenderer();
			TableCellRenderer rendCol = tableColumn.getHeaderRenderer();
		    if (rendCol == null) rendCol = rend;
		    Component header = rendCol.getTableCellRendererComponent(table, tableColumn.getHeaderValue(), false, false, 0, col);
		    int maxWidth = header.getPreferredSize().width + 15;
		    w = Math.max(w, maxWidth);

			for (int row = 0; row < table.getRowCount(); row++) {
				TableCellRenderer tableCellRenderer = table.getCellRenderer(row, col);
				Component c = table.prepareRenderer(tableCellRenderer, row, col);
				int width = c.getPreferredSize().width + table.getIntercellSpacing().width + 15;
				preferredWidth = Math.max(width, preferredWidth);

		        if (preferredWidth <= maxWidth){
			        // Exceeded the maximum width, no need to check other rows
		            preferredWidth = maxWidth;
		            break;
		        }
			}

			preferredWidth = Math.max(w, preferredWidth);
			tableColumn.setPreferredWidth(preferredWidth);
		}
	}
```

Seems like we only used it once when initializing the table. In that case, the Historical Event tab is initially empty with zero incidents and the column width would NOT be computed correctly.

So, where is the best place to call adjustColumnWidth() periodically or on as-needed basis ?

**Comment by bevans2000** [comment]
> 



> I see quite a lot of this log on a vehicle.
> 
> ```
> 01-Adir-02:788.701 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:788.701 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:788.846 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:788.846 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:788.991 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:788.991 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:789.136 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:789.136 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:789.292 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:789.292 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:789.443 (Warning) EntityTableModel : Entity not in model : Martian Escape
> 01-Adir-02:789.443 (Warning) EntityTableModel : Entity not in model 
> ```
> 
> Can you take a quick look ?
Sure. If its annoying must comment out the warning message. It was a check message I needed during the development.
Did the message go away when you changed tab or the selected Settlement?
It means there is a Unit.Listener in place for a Vehicle not displayed. In the old code this silently happened all the time but I thought I had covered it.

**Comment by bevans2000** [comment]
I would only call the adjuster once per TableTab but I don't see why the EventTable would be zero without it. Because it has  o rows?
DL you want the ability to predefined the column widths to remove those altogether?

**Comment by bevans2000** [comment]
> > Can you take a quick look ?
> > Sure. If its annoying must comment out the warning message. It was a check message I needed during the development.
> > Did the message go away when you changed tab or the selected Settlement?
> > It means there is a UnitListener in place for a Vehicle not displayed. In the old code this silently happened all the time but I thought I had covered it.

I have push a commit to remove the message but I know what it is. It is nothing wrong but the why the existing code uses a ```SwingUtilities``` to update the model from the UnitListener. By time the event is executed the Unit may have been removed  if the Settlement filter is changed.

**Comment by bevans2000** [comment]
I will tidy up how the various model flag a changed cell. There is a mixture of approaches, some synch and some asynch.  Can push the logic into EntityTableModel so we control it from one place.

**Comment by mokun** [comment]
> some synch and some asynch

which one would you consider synch and which one asynch ?

**Comment by mokun** [comment]
There's NPE and ConcurrentModificaionException from time to time regarding EntityTableModel.

For instance,

```
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
	at org.mars_sim.msp.core.person.Person.getAmountResourceStored(Person.java:1935)
	at org.mars_sim.msp.core.structure.Settlement.getAllAmountResourceOwned(Settlement.java:3937)
	at org.mars_sim.msp.ui.swing.tool.monitor.SettlementTableModel.getEntityValue(SettlementTableModel.java:290)
	at org.mars_sim.msp.ui.swing.tool.monitor.SettlementTableModel.getEntityValue(SettlementTableModel.java:1)
	at org.mars_sim.msp.ui.swing.tool.monitor.EntityTableModel.getValueAt(EntityTableModel.java:268)
	at org.mars_sim.msp.ui.swing.tool.monitor.TableSorter.getValueAt(TableSorter.java:203)
	at java.desktop/javax.swing.JTable.getValueAt(JTable.java:2706)
	at java.desktop/javax.swing.JTable.prepareRenderer(JTable.java:5724)
	at com.alee.laf.table.TablePainter.paintCell(TablePainter.java:740)
	at com.alee.laf.table.TablePainter.paintContent(TablePainter.java:655)
	at com.alee.laf.table.TablePainter.paintContent(TablePainter.java:402)
	at com.alee.laf.table.TablePainter.paintContent(TablePainter.java:38)
	at com.alee.painter.decoration.AbstractDecorationPainter.paint(AbstractDecorationPainter.java:1363)
	at com.alee.painter.PainterSupport.paint(PainterSupport.java:557)
	at com.alee.laf.table.WebTableUI.paint(WebTableUI.java:174)
	at java.desktop/javax.swing.plaf.ComponentUI.update(ComponentUI.java:161)
	at java.desktop/javax.swing.JComponent.paintComponent(JComponent.java:797)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1074)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JViewport.paint(JViewport.java:737)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JLayeredPane.paint(JLayeredPane.java:590)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JLayeredPane.paint(JLayeredPane.java:590)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JLayer.paint(JLayer.java:475)
	at java.desktop/javax.swing.plaf.LayerUI.paint(LayerUI.java:80)
	at org.mars_sim.msp.ui.swing.tool.WaitLayerUIPanel.paint(WaitLayerUIPanel.java:44)
	at java.desktop/javax.swing.plaf.ComponentUI.update(ComponentUI.java:161)
	at java.desktop/javax.swing.JComponent.paintComponent(JComponent.java:797)
	at java.desktop/javax.swing.JLayer.paint(JLayer.java:470)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JComponent.paintToOffscreen(JComponent.java:5255)
	at java.desktop/javax.swing.RepaintManager$PaintManager.paintDoubleBufferedImpl(RepaintManager.java:1643)
	at java.desktop/javax.swing.RepaintManager$PaintManager.paintDoubleBuffered(RepaintManager.java:1618)
	at java.desktop/javax.swing.RepaintManager$PaintManager.paint(RepaintManager.java:1556)
	at java.desktop/javax.swing.RepaintManager.paint(RepaintManager.java:1323)
	at java.desktop/javax.swing.JComponent._paintImmediately(JComponent.java:5203)
	at java.desktop/javax.swing.JComponent.paintImmediately(JComponent.java:5013)
	at java.desktop/javax.swing.JLayer.paintImmediately(JLayer.java:430)
	at java.desktop/javax.swing.plaf.LayerUI.paintImmediately(LayerUI.java:723)
	at java.desktop/javax.swing.JLayer.paintImmediately(JLayer.java:425)
	at java.desktop/javax.swing.JComponent.paintImmediately(JComponent.java:4994)
	at java.desktop/javax.swing.RepaintManager$4.run(RepaintManager.java:865)
	at java.desktop/javax.swing.RepaintManager$4.run(RepaintManager.java:848)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:848)
	at java.desktop/javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:823)
	at java.desktop/javax.swing.RepaintManager.prePaintDirtyRegions(RepaintManager.java:772)
	at java.desktop/javax.swing.RepaintManager$ProcessingRunnable.run(RepaintManager.java:1890)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

**Comment by bevans2000** [comment]
> > some synch and some asynch
> 
> which one would you consider synch and which one asynch ?

If we use synch then it removes the change of the  I current modification exception as the Unit Thread is actually doing the table repaint but it impacts on the tick performance.
If we async then we isolate the Unit Thread but increase the chance of concurrency problems.
However I think we can do something clever that adapts if these derived attributes are marked as cached in the table model. We must make sure than any derived attribute is cached, then we refresh the cache  as part on the listener logic instead instead the paint logic. Thus the goal is the table repaint logic only uses value already in memory that do not need to be recalculated. The actual UI painting consumes a lot of processing.
I think this will be better but needs a but if a change. I will look into this, I assume you are running with a large number of Settlements?

**Comment by bevans2000** [comment]
OK, so I've made the change handle async events. The keys is that any column that involves an expensive calculation should be marked as cached. These will be calculated when the event fires and hence in the Thread of the UnitManager in theory avoiding any Threading problems.

**Comment by mokun** [comment]
After this recent change, the resource amounts in Mars tab or other settlement's tab (using SettlementTableModel) are in 2 decimal places again.

If they are in integer, the change in value would be a little less frequent, right ?

**Comment by bevans2000** [comment]
So I tried a little trick here........ If the column is defined as returning a Number then the renderer is to 2 decimal places. If it is defined as a Double type then it is to 3 decimal places. This allows the model classes to influence the layout with needing to define cell renderers for each table.
So depending on the column type in the model class we can control how accurate the render is.

**Comment by mokun** [comment]
Good job patching the Swing UI but guess it's about time we switch to JavaFX, shall we ?

**Comment by bevans2000** [comment]
Thanks. It's been a long time since I touched any Swing code. 

As to JavaFX, one big drawback is it's not bundled with Java. The current client works for me but if you want to move this way then fine. For me the long term goals it to convert it to a web application.
